src/webgpu/api/validation/encoding/cmds/debug.spec.ts

```diff
@@ -0,0 +1,143 @@
+  GPURenderBundleEncoder):
+    - Test that all pushDebugGroup must have a corresponding popDebugGroup
+      - Push and pop counts of 0, 1, and 2 will be used.
+      - An error must be generated for non matching counts.
+    - Test calling pushDebugGroup with an empty string.
+    - Test inserting a debug marker with an empty string.
+    - Test inserting a debug marker with a non-empty string.
+`;
+
+import { poptions, params } from '../../../../../common/framework/params_builder.js';
+import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { assert } from '../../../../../common/framework/util/util.js';
+
+import { ValidationTest } from './../../validation_test.js';
+
+type Encoders = GPUCommandEncoder | GPUProgrammablePassEncoder;
+const kEncoderTypes = ['non-pass', 'compute pass', 'render pass', 'render bundle'] as const;
+type EncoderTypes = typeof kEncoderTypes[number];
+
+class F extends ValidationTest {
+  #commandEncoder: GPUCommandEncoder | undefined = undefined;
+
+  makeAttachmentTexture(): GPUTexture {
+    return this.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 16, height: 16, depth: 1 },
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+  }
+
+  createEncoder(encoderType: EncoderTypes): Encoders {
+    assert(this.#commandEncoder === undefined);
+    switch (encoderType) {
+      case 'non-pass':
+        return this.device.createCommandEncoder({});
+      case 'render bundle':
+        return this.device.createRenderBundleEncoder({ colorFormats: ['rgba8unorm'] });
+      case 'compute pass':
+        this.#commandEncoder = this.device.createCommandEncoder({});
+        return this.#commandEncoder.beginComputePass({});
+      case 'render pass':
+        this.#commandEncoder = this.device.createCommandEncoder({});
+        return this.#commandEncoder.beginRenderPass({
+          colorAttachments: [
+            {
+              attachment: this.makeAttachmentTexture().createView(),
+              loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+            },
+          ],
+        });
+    }
+  }
+
+  finishEncoder(encoder: Encoders, encoderType: EncoderTypes) {
+    let commandBuffer: GPUCommandBuffer | undefined = undefined;
+    switch (encoderType) {
+      case 'non-pass': {
+        commandBuffer = (encoder as GPUCommandEncoder).finish({});
+        break;
+      }
+      case 'render bundle': {
+        const bundle = (encoder as GPURenderBundleEncoder).finish({});
+        const commandEncoder = this.device.createCommandEncoder({});
+        const pass = commandEncoder.beginRenderPass({
+          colorAttachments: [
+            {
+              attachment: this.makeAttachmentTexture().createView(),
+              loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+            },
+          ],
+        });
+        pass.executeBundles([bundle]);
+        pass.endPass();
+        commandBuffer = commandEncoder.finish({});
```

Comment:
The abstraction you've built here over renderPass and renderBundle is nice and looks very useful. I think it could be useful for other tests as well (we might even have some already that try to do a similar abstraction). Perhaps as a class that creates either a pass or a bundle, exposes it as a GPURenderEncoderBase, and has a finish() method similar to this which returns a GPUCommandBuffer. Perhaps with extended versions for GPUProgrammablePassEncoder (same + compute pass) and `GPUCommandEncoder | GPUProgrammablePassEncoder` (same as what you have here).

Would you be interested in looking at this as a follow-up patch? (Otherwise I'll put it on my infinite todo list.)

Reply:
Sure, I can take a look at it.